src/phase0/state_transition/mod.rs

```diff
@@ -0,0 +1,387 @@
+}
+
+pub fn get_domain<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    domain_type: DomainType,
+    epoch: Option<Epoch>,
+) -> Result<Domain, Error> {
+    let epoch = if epoch.is_none() {
+        get_current_epoch(&state)
+    } else {
+        epoch.unwrap()
+    };
+    let fork_version = if epoch < state.fork.epoch {
+        Some(state.fork.previous_version.clone())
+    } else {
+        Some(state.fork.current_version.clone())
+    };
+
+    compute_domain(
+        domain_type,
+        fork_version,
+        Some(state.genesis_validators_root.clone()),
+    )
+}
+
+pub fn get_current_epoch<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+) -> Epoch {
+    compute_epoch_at_slot(state.slot)
+}
+
+pub fn compute_signing_root<T: SimpleSerialize>(
+    ssz_object: &T,
+    domain: Domain,
+) -> Result<Root, Error> {
+    let context = MerkleizationContext::new();
+    let object_root = ssz_object.hash_tree_root(&context)?;
+
+    let s = SigningData {
+        object_root,
+        domain,
+    };
+    s.hash_tree_root(&context)
+        .map_err(Error::MerkleizationError)
+}
+
+pub fn compute_epoch_at_slot(slot: Slot) -> Epoch {
+    slot / SLOTS_PER_EPOCH
+}
+
+pub fn compute_start_slot_at_epoch(epoch: Epoch) -> Slot {
+    epoch * SLOTS_PER_EPOCH
+}
+
+pub fn compute_activation_exit_epoch(epoch: Epoch) -> Epoch {
+    epoch + 1 + MAX_SEED_LOOKAHEAD
+}
+
+pub fn compute_fork_digest(
+    current_version: Version,
+    genesis_validators_root: Root,
+) -> Result<ForkDigest, Error> {
+    let fork_data_root = compute_fork_data_root(current_version, genesis_validators_root)?;
+    let fork_digest: Vector<u8, 4> = ForkDigest::deserialize(fork_data_root.as_ref())?;
```

Comment:
using `deserialize` here should actually always error bc the input will be longer than the possible length of 4 bytes; and also we know the root is in fact a `Root` so no need to deserialize, which in general may be some strange pathological input that needs special care to internalize.

so instead, i think here we can do something like:
```rust
let fork_digest = fork_data_root[..4].try_into().expect("is the correct length");
```

where we want to ensure that a `ForkDigest` is just an alias for `[u8;4]`. the latest version of `ssz_rs` should have support for arrays of bytes like this.

Reply:
yes, it works with the latest ssz_rs. I was not aware of the changes.
Fixed.

Reply:
It doesn't work